### PR TITLE
Feature TCK - initial

### PR DESCRIPTION
**Pull request by stbischof**
Currently this is not the newest version of the Apache sling feature-service implementation.
Will change this to a maven-snapshot as soon as available.

### CODE REVIEW

**Comment by bosschaert** [comment]
👍

**Comment by stbischof** [comment]
@bjhargrave are you able to return the job?

**Comment by bjhargrave** [comment]
> @bjhargrave are you able to return the job?

The TCK execution failed because of a resolve error. You should fix that.

**Comment by stbischof** [comment]
i just saw 
` Error: We are currently unable to download the log. Please try again later. `

**Comment by stbischof** [comment]
After https://github.com/apache/felix-dev/pull/88 is now merged. This should generate a snapshot that i will use

File: org.osgi.test.cases.feature/src/org/osgi/test/cases/feature/junit/FeatureIDTest.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+			"g:a", "g:a:", "g:a::", "g:a:::", //
+			"g:a:v:", "g:a:v::", //
+			"g:a:t:c:", //
+			"g:a:t::v", //
+			"g::t:c:v", //
+			"g:::t:v", //
+			":a:t:c:v", //
+			"g:a:t:c:v:", //
+			"g:a:t:c:v:foo", //
+			"*:a:v", "g:*:v", "g:a:*", "g:a:*:*:v", //
```
[the remark this answers is not in the document]
Comment (stbischof) [reply org.osgi.test.cases.feature/src/org/osgi/test/cases/feature/junit/FeatureIDTest.java:177]: agree

File: org.osgi.test.cases.feature/src/org/osgi/test/cases/feature/junit/FeatureIDTest.java [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+				.hasEmptyType()
+				.hasEmptyClassifier()
+				.hasToString(G_A_V);
+
+	}
+
+	@Test
+	void constructor5_gav() throws Exception {
+
+		ID id = fs.getID("g", "a", "v", null, null);
```
[the remark this answers is not in the document]
Comment (stbischof) [reply org.osgi.test.cases.feature/src/org/osgi/test/cases/feature/junit/FeatureIDTest.java:48]: removed

File: org.osgi.test.cases.feature/src/org/osgi/test/cases/feature/junit/FeatureIDTest.java [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+				.hasEmptyType()
+				.hasEmptyClassifier()
+				.hasToString(G_A_V);
+
+	}
+
+	@Test
+	void getID5_gavt() throws Exception {
+
+		ID id = fs.getID("g", "a", "v", "t", null);
```
[the remark this answers is not in the document]
Comment (stbischof) [reply org.osgi.test.cases.feature/src/org/osgi/test/cases/feature/junit/FeatureIDTest.java:61]: switched to test the 4 arg construcktor

**Comment by stbischof** [comment]
**Open Questions:**

* is this a valid  Configuration JSON?

```
{
    // Resource Format Version
    ":configurator:resource-version" : 1,

    // Simple Configuration
    "pid.a": {
           "key": "val"
        },

    // Configuration with Prop-Null
    "pid.b": null
}
```

* Should OSGI-API be packed into the Implementation?

**Comment by bosschaert** [comment]
If at this stage we just have a few failing tests left, maybe we should merge this branch. 
We can discuss the open items during the spec call of course, but I think it would be nice if we could merge this branch now soon. WDYT @bjhargrave ?

**Comment by bjhargrave** [comment]
> WDYT @bjhargrave ?

That is fine with me. Lets merge.

File: org.osgi.test.cases.feature/.classpath [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8">
+		<attributes>
+			<attribute name="module" value="true"/>
```
Comment (bjhargrave) [inline org.osgi.test.cases.feature/.classpath:5]: What is this attribute? We don't have it in any of the other .classpath files.

File: org.osgi.test.cases.feature/.classpath [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+		</attributes>
+	</classpathentry>
+	<classpathentry kind="con" path="aQute.bnd.classpath.container"/>
+	<classpathentry kind="src" path="src"/>
+	<classpathentry kind="src" path="test"/>
```
Comment (bjhargrave) [inline org.osgi.test.cases.feature/.classpath:10]: There is no test folder. Remove this line.

File: cnf/ext/central.mvn [hunk 6]
```diff
@@ -197,4 +197,7 @@ org.apache.zookeeper:zookeeper:3.4.13
 org.apache.aries.component-dsl:org.apache.aries.component-dsl.component-dsl:1.2.2
 org.apache.aries.typedevent:org.apache.aries.typedevent.bus:0.0.1-SNAPSHOT
 
-
+org.apache.felix:org.apache.felix.feature:0.0.1-SNAPSHOT
```
Comment (bjhargrave) [inline cnf/ext/central.mvn:200]: Isn't there already an RC1 version in maven central? (RC2 is under ballot.)

**Review by bjhargrave — CHANGES_REQUESTED** [review]

**Comment by bjhargrave** [comment]
You should also rebase on main since the add of mockito was made in the typed events tck PR.

File: org.osgi.test.cases.feature/bnd.bnd [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+Import-Package: ${-signaturetest}, *
+
+-buildpath: \
+	org.osgi.test.support;version=project,\
+	org.osgi.service.feature;version=latest,\
+	osgi.core;maven-scope=provided;version='4.3.1',\
+	org.json;version='1.0.0',\
+	org.apache.geronimo.specs.geronimo-json_1.1_spec
+
+-dependson: org.osgi.service.feature
```
[the remark this answers is not in the document]
Comment (stbischof) [reply org.osgi.test.cases.feature/bnd.bnd:16]: removed

File: org.osgi.test.cases.feature/bnd.bnd [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+-runbundles						=\
+	org.osgi.impl.bundle.serviceloader;version=latest,\
+	org.osgi.impl.service.feature; version=latest,\
+	org.osgi.service.feature; version=latest,\
+	org.osgi.impl.service.component; version=latest,\
+	org.osgi.util.converter; version=latest,\
+	org.mockito.mockito-core
```
[the remark this answers is not in the document]
Comment (stbischof) [reply org.osgi.test.cases.feature/bnd.bnd:23]: removed

File: org.osgi.test.cases.feature/.classpath (hunk 4 above)
Comment (stbischof) [reply org.osgi.test.cases.feature/.classpath:5]: uh.. no idea where this came from

File: cnf/ext/central.mvn (hunk 6 above)
Comment (stbischof) [reply cnf/ext/central.mvn:200]: i use this one 
https://repository.apache.org/content/groups/snapshots/org/apache/felix/org.apache.felix.feature/0.9.1-SNAPSHOT/

File: org.osgi.test.cases.feature/bnd.bnd [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+	org.osgi.test.support;version=project,\
+	org.osgi.service.feature;version=latest,\
+	osgi.core;maven-scope=provided;version='4.3.1',\
+	org.json;version='1.0.0',\
+	org.apache.geronimo.specs.geronimo-json_1.1_spec
+
+-runbundles						=\
+	org.osgi.impl.bundle.serviceloader;version=latest,\
+	org.osgi.impl.service.feature; version=latest,\
+	org.osgi.service.feature; version=latest,\
```
Comment (bjhargrave) [inline org.osgi.test.cases.feature/bnd.bnd:19]: You should move this line into the lib file.

Comment (stbischof) [reply org.osgi.test.cases.feature/bnd.bnd]: done

**Comment by stbischof** [comment]
Waiting for the RC3 release/ tested against the local build snapshot of felix.feature

**Review by bjhargrave — APPROVED** [review]

**Pull request merged**